src/query/functions/temporal/base.go

```diff
@@ -0,0 +1,505 @@
+}
+
+// MakeProcessor is a way to create a transform
+type MakeProcessor func(op baseOp, controller *transform.Controller) Processor
+
+type processRequest struct {
+	blk    block.Block
+	bounds block.Bounds
+	deps   []block.Block
+}
+
+// blockCache keeps track of blocks from the same parent across time
+type blockCache struct {
+	mu              sync.Mutex
+	initialized     bool
+	blockList       []block.Block
+	op              baseOp
+	transformOpts   transform.Options
+	startBounds     block.Bounds
+	endBounds       block.Bounds
+	processedBlocks []bool
+}
+
+func newBlockCache(op baseOp, transformOpts transform.Options) *blockCache {
+	return &blockCache{
+		op:            op,
+		transformOpts: transformOpts,
+	}
+}
+func (c *blockCache) init(bounds block.Bounds) {
+	if c.initialized {
+		return
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.initialized {
+		return
+	}
+
+	timeSpec := c.transformOpts.TimeSpec
+	c.startBounds = bounds.Nearest(timeSpec.Start)
+	c.endBounds = bounds.Nearest(timeSpec.End.Add(-1 * bounds.StepSize))
+	numBlocks := c.endBounds.End().Sub(c.startBounds.Start) / bounds.Duration
+	c.blockList = make([]block.Block, numBlocks)
+	c.processedBlocks = make([]bool, numBlocks)
+	c.initialized = true
+}
+
+func (c *blockCache) index(t time.Time) (int, error) {
+	start := c.startBounds.Start
+	if t.Before(start) || t.After(c.endBounds.Start) {
+		return 0, fmt.Errorf("invalid time for the block cache: %v, start: %v, end: %v", t, start, c.endBounds.Start)
+	}
+
+	return int(t.Sub(start) / c.startBounds.Duration), nil
+}
+
+// Add the block to the cache, errors out if block already exists
+func (c *blockCache) add(key time.Time, b block.Block) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	index, err := c.index(key)
+	if err != nil {
+		return err
+	}
+
+	if c.blockList[index] != nil {
+		return fmt.Errorf("block already exists at index: %d", index)
+	}
+
+	c.blockList[index] = b
+	return nil
+}
+
+// Remove the block from the cache
+func (c *blockCache) remove(idx int) error {
+	if idx >= len(c.blockList) {
+		return fmt.Errorf("index out of range for remove: %d", idx)
+	}
+
+	c.mu.Lock()
+	c.blockList[idx] = nil
+	c.mu.Unlock()
+
+	return nil
+}
+
+// Get the block from the cache
+func (c *blockCache) get(key time.Time) (block.Block, bool) {
+	c.mu.Lock()
+	index, err := c.index(key)
+	if err != nil {
+		c.mu.Unlock()
+		return nil, false
+	}
+
+	b := c.blockList[index]
+	c.mu.Unlock()
+	return b, b != nil
+}
+
+// multiGet retrieves multiple blocks from the cache at once until if finds an empty block
+func (c *blockCache) multiGet(startBounds block.Bounds, numBlocks int, reverse bool) ([]block.Block, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	blks := make([]block.Block, 0, numBlocks)
+	if numBlocks == 0 {
+		return blks, nil
+	}
+
+	startIdx, err := c.index(startBounds.Start)
+	if err != nil {
+		return nil, err
+	}
+
+	// Process a single index
+	process := func(i int) (bool, error) {
```

Comment:
nit: rename to addIfPresent, maybe add a comment as to what this does. Also, flip to return false if not added, and true if added, since true usually corresponds to the positive case

Reply:
i made the name clearer and added a comment. I still think its reads better with empty, let me know once you read the new comment/naming